### PR TITLE
Refactor UploadRepo and add tests

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
In doing this, I found an issue with the existing upload code. Unfortunately, it's a bit buried, but the S3 upload session won't ever request a new upload session because the expiration was going the wrong way. Found that during testing.

This is still *not* including any actual changes to the iOS uploading.

### CODE REVIEW

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt [hunk 1]
```diff
@@ -113,7 +156,7 @@ internal open class UploadRepo(
                     TimeZone.currentSystemDefault()
                 )
                 val expires = Instant.parse(uploadSession.expires)
-                if (expires < desiredMinimumExpiration) {
+                if (desiredMinimumExpiration < expires) {
```
Comment (syoung-smallwisdom) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt:159]: This is the line that was failing the unit test. Should only return the cached upload session if upload session will be valid for more than 30 minutes before it expires.

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt [hunk 2]
```diff
@@ -5,26 +5,17 @@ import kotlinx.serialization.Serializable
 
 @Serializable
 data class UploadFile (
-    val filePath: String,
+    override val filePath: String,
     val contentType: String,
     val fileLength: Long,
-    val md5Hash: String,
+    val md5Hash: String?,
```
Comment (nategbrown9) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt:11]: Is there a reason this needs to be nullable?

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt [hunk 3]
```diff
@@ -140,33 +183,30 @@ internal open class UploadRepo(
...
-    suspend fun processFinishedUploads() : Boolean {
-        //Check that all uploads succeeded and have been removed
-        val success = database.getResources(ResourceType.FILE_UPLOAD,
+    /**
+     * Process any [UploadSession] records where the upload to S3 has been completed, but Bridge
+     * has not been notified.
+     */
+    suspend fun processFinishedUploads() {
+        // Check if we have any sessions that have been uploaded and still need completing
+        for (resource in database.getResourcesNeedSave(
```
Comment (nategbrown9) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt:203]: UploadSession will need to have needSave set to true when written to cache for this to work.

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt (hunk 2 above)
Comment (syoung-smallwisdom) [reply bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt:11]: There are two paths for uploading data to S3. Currently, only "Study Data" is supported via BridgeClient. The other is "Participant Data" and the headers for it do not include the MD5 hash. I stubbed out using the `UploadFile` and the `S3UploadSession` model object to allow supporting participant data at a future time.

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt [hunk 4]
```diff
@@ -67,31 +72,69 @@ internal open class UploadRepo(
...
-            lastUpdate = Clock.System.now().toEpochMilliseconds(),
-            status = ResourceStatus.SUCCESS,
-            needSave = false
-        )
-        database.insertUpdateResource(resource)
+    fun markUploadFileFinished(uploadFile: UploadFileIdentifiable) {
+        val resourceId = uploadFile.getUploadSessionResourceId()
+        val resource = database.getResource(resourceId, ResourceType.UPLOAD_SESSION,
+            ResourceDatabaseHelper.APP_WIDE_STUDY_ID
+        )?.copy(needSave = true)
```
Comment (syoung-smallwisdom) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt:87]: @nategbrown9 
When the S3 upload is finished, the `UploadFile` resource is deleted and the `UploadSession` resource is marked as `needSave = true`.

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt (hunk 3 above)
Comment (syoung-smallwisdom) [reply bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt:203]: That's handled in `fun markUploadFileFinished(uploadFile: UploadFileIdentifiable)` where the `UploadSession` is set to needSave. This allows for a more accurate count of how many pending uploads there are.

File: bridge-client/src/commonTest/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepoTest.kt [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+            assertEquals(expectedS3UploadSession, response)
+
+            val initialCount = database.getPendingUploadCount()
+            assertEquals(1L, initialCount)
+
+            // Pretend to upload the S3 file by marking it as finished
+            repo.markUploadFileFinished(UploadFileId(tempFile))
+
+            val afterUploadCount = database.getPendingUploadCount()
+            assertEquals(1L, afterUploadCount)
```
Comment (syoung-smallwisdom) [inline bridge-client/src/commonTest/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepoTest.kt:145]: This check is here to make sure that both before and after marking the file as uploaded, the pending count is still == 1.

File: bridge-client/src/commonTest/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepoTest.kt [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+            // Finally process finished uploads
+            repo.processFinishedUploads()
+
+            assertTrue(database.getResources(ResourceType.FILE_UPLOAD, ResourceDatabaseHelper.APP_WIDE_STUDY_ID).isEmpty())
+            assertTrue(database.getResources(ResourceType.UPLOAD_SESSION, ResourceDatabaseHelper.APP_WIDE_STUDY_ID).isEmpty())
+
+            val uploadsAllFinishedCount = database.getPendingUploadCount()
+            assertEquals(0L, uploadsAllFinishedCount)
```
Comment (syoung-smallwisdom) [inline bridge-client/src/commonTest/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepoTest.kt:154]: But after the uploads are processed, the pending upload count drops to 0.

**Review by nategbrown9 — APPROVED** [review]
Looks good.

**Pull request merged**